spring-kafka/src/main/java/org/springframework/kafka/annotation/RetryTopicRegistryPostProcessor.java

```diff
@@ -0,0 +1,112 @@
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.annotation;
+
+import java.util.Arrays;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.beans.factory.support.BeanDefinitionRegistry;
+import org.springframework.beans.factory.support.BeanDefinitionRegistryPostProcessor;
+import org.springframework.beans.factory.support.RootBeanDefinition;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.ApplicationContextAware;
+import org.springframework.core.Ordered;
+import org.springframework.core.type.AnnotationMetadata;
+import org.springframework.kafka.retrytopic.RetryTopicBootstrapper;
+import org.springframework.kafka.retrytopic.RetryTopicConfiguration;
+import org.springframework.kafka.retrytopic.RetryTopicInternalBeanNames;
+
+/**
+ * A {@link BeanDefinitionRegistryPostProcessor} implementation that registers
+ * the non-blocking delayed retries feature's components' beans if a
+ * {@link RetryableTopic} annotated method or a {@link RetryTopicConfiguration}
+ * bean is found.
+ *
+ * @author Tomaz Fernandes
+ * @since 2.9.0
+ *
+ * @see RetryTopicBootstrapper
+ * @see RetryTopicConfiguration
+ * @see RetryableTopic
+ */
+public class RetryTopicRegistryPostProcessor
+		implements BeanDefinitionRegistryPostProcessor, Ordered, ApplicationContextAware {
+
+	private static final String RETRY_TOPIC_BOOTSTRAPPER_BEAN_NAME = RetryTopicInternalBeanNames
+			.RETRY_TOPIC_BOOTSTRAPPER;
+	static final String RETRY_TOPIC_BOOTSTRAPPED_ATTRIBUTE = "retryTopicBootstrapped";
+	private ApplicationContext applicationContext;
+
+	@Override
+	public void postProcessBeanDefinitionRegistry(BeanDefinitionRegistry registry) throws BeansException {
+		if (registry.containsBeanDefinition(RetryTopicInternalBeanNames
+				.RETRY_TOPIC_BOOTSTRAPPER) && registry.getBeanDefinition(RetryTopicInternalBeanNames
+				.RETRY_TOPIC_BOOTSTRAPPER).hasAttribute(RETRY_TOPIC_BOOTSTRAPPED_ATTRIBUTE)) {
+
+			return;
+		}
+
+		Arrays
+				.stream(registry.getBeanDefinitionNames())
+				.map(registry::getBeanDefinition)
+				.filter(this::hasRetryTopicConfiguration)
+				.findFirst()
+				.ifPresent(beanDef -> bootstrapRetryTopic(registry));
+	}
+
+	private boolean hasRetryTopicConfiguration(BeanDefinition beanDef) {
+		return RetryTopicConfiguration.class.isAssignableFrom(beanDef.getResolvableType().toClass())
+				|| AnnotationMetadata
+					.introspect(beanDef.getResolvableType().toClass())
+					.hasAnnotatedMethods(RetryableTopic.class.getName());
+	}
+
+	private void bootstrapRetryTopic(BeanDefinitionRegistry registry) {
+		BeanDefinition beanDef = getOrRegisterBeanDefinition(registry);
+		this.applicationContext.getBean(RETRY_TOPIC_BOOTSTRAPPER_BEAN_NAME, RetryTopicBootstrapper.class)
+				.bootstrapRetryTopic();
```

Comment:
This better to be done via `SmartInitializingSingleton`  impl on that `RetryTopicBootstrapper`. So, we don't need to have this `getBean()` over here and no early `BeanFactory` access.
Said that it feels like we don't need this `RetryTopicRegistryPostProcessor` at all...

Reply:
Again, I think it's about the premise. If we want to keep the component's bootstrapping optional while also enabling their autowiring to the end user, I think the `SmartInitializingSingleton` hook is too late for that.

Reply:
Do we really need to have them to be "autowireable" in the end?
Are they not infrastructure beans to support something else which is already those `@RetryableTopic` or so?

Reply:
> OK. All sounds good.
> 
> Here is my take on this: let's make those beans registered unconditionally if there is a requirements to be able to inject them somewhere in the end-user code. Even if these beans are going to be dangling for nothing, doesn't look like they are heavy, active or store some context to exhaust memory or CPU.
> 
> We may revise a conditional feature in the future.

Now that I think of it, we have two kinds of components - infrastructure components used at assembly time, and runtime components that are used at, well, runtime.

I think the runtime components are the ones users might be interested in. For example in #2172, the `DestinationTopicContainer` is injected for the user to figure out what is the first retry-topic name for a given main topic and reprocess DLT messages.

I think it might also be useful for users to have access to the `KafkaConsumerBackOffManager` so they can know which topic partitions are currently backed off and their due time, for example. But that one might be trickier because it has a more complex instantiation, but that's something fixable, and probably we won't need it now.

So maybe we can register early only `DestinationTopicContainer` and maybe `KafkaBackOffManager` and let the infrastructure components be instantiated where they currently are, when `KLABPP` detects the feature.

How does this sound? Would `KafkaBootstrapConfiguration` be the right place for that?

> 
> Your scan for `@RetryableTopic` doesn't look bad. Only the problem here that we need opposite injection from these infrastructure beans into end-user stuff. This is kinda anti-pattern and hard to implement and support eventually.

I'm not sure I understand what opposite injection is. You mean restricting users' access to the infrastrucure beans via injection?

Reply:
Yes. Sounds like a plan.

I mean that you are going to register a bean dynamically, but at the same time it is asked for end-user autowiring. 
That's my concern where we may cause circular dependency.
That's where it might not be conditional. At least for now.

Reply:
> Yes. Sounds like a plan.
> 

Great, I'll do that then. I'll also take a closer look if there are any other beans that might be interesting for users.

> I mean that you are going to register a bean dynamically, but at the same time it is asked for end-user autowiring. That's my concern where we may cause circular dependency. That's where it might not be conditional. At least for now.

Oh, I see, thanks for the explanation.

Thanks a lot for looking into this!